### ISSUE TITLE
How to compute similarity score between different modalities?

### ISSUE THREAD

**Issue by alparius** [invictus717/MetaTransformer#42]
Hello! The project looks very promising, but I'm having some issues starting up with it. Given that there is a common embedding space for all these modalities, all I would like to do with it is to encode and compare different data types.

Very rudimentary example would be that I have some pictures of animals and some point clouds of animals and I'd like to calculate the similarity matrix to find out which picture matches which point cloud.

As far as I understand it,

1. this [README.md / Model Zoo](https://github.com/invictus717/MetaTransformer#-model-zoo) snippet loads the foundation model
2. then I'd need to encode all my different file types with some functions that I can't find - I'm stuck here
3. multiply, maybe softmax etc. the two feature arrays to get a similarity matrix/heatmap

**Comment by 1223haohao** [comment]
> Hello! The project looks very promising, but I'm having some issues starting up with it. Given that there is a common embedding space for all these modalities, all I would like to do with it is to encode and compare different data types.
> 
> Very rudimentary example would be that I have some pictures of animals and some point clouds of animals and I'd like to calculate the similarity matrix to find out which picture matches which point cloud.
> 
> As far as I understand it,
> 
> 1. this [README.md / Model Zoo](https://github.com/invictus717/MetaTransformer#-model-zoo) snippet loads the foundation model
> 2. then I'd need to encode all my different file types with some functions that I can't find - I'm stuck here
> 3. multiply, maybe softmax etc. the two feature arrays to get a similarity matrix/heatmap

I want to express my sincere appreciation to the authors for creating such a promising piece of work. 

I have a similar question. I'm struggling to comprehend how to map data from different modalities into tokens. From what I understand, this process is handled within the 'Data2Seq' module. However, it appears to be implemented in a straightforward Python file.

Does this imply that I need to pre-train a tokenizer before running the code? The Python file within the 'Data2Seq' module is primarily for demonstration purposes and isn't intended for direct user execution.

**Comment by invictus717** [comment]
Thank you for your interest in Meta-Transformer! 

* Pretrained Tokenizers?

You can load pretrained tokenizers with [MAE](https://github.com/facebookresearch/mae) or other pretraining methods. We've tested that they are compatible.

I'm willing to provide some demo implementation as follows, which I hope will help:

```python

From Data2Seq import Data2Seq

video_tokenier = Data2Seq(modality='video',dim=768)

audio_tokenier = Data2Seq(modality='audio',dim=768)

time_series_tokenier = Data2Seq(modality='time-series',dim=768)

features = torch.concat([video_tokenizer(video), audio_tokenizer(audio), time_series_tokenizer(time_data)],dim=1)

from timm.models.vision_transformer import Block
ckpt = torch.load("Meta-Transformer_base_patch16_encoder.pth")
encoder = nn.Sequential(*[
            Block(
                dim=768,
                num_heads=12,
                mlp_ratio=4.,
                qkv_bias=True,
                norm_layer=nn.LayerNorm,
                act_layer=nn.GELU
            )
            for i in range(12)])
encoder.load_state_dict(ckpt,strict=True)
encoded_features = encoder(features)
```

As for the similarity scores, you can simply compute as follows:
```python
feat_1 = encoded_features[0]
feat_2 = encoded_features[1]
sim = feat_1 @ feat_2.transpose(-2, -1)
```
